modules/basics/src/main/java/com/opengamma/strata/basics/currency/FxRate.java

```diff
@@ -236,6 +237,23 @@ public FxRate toConventional() {
 
   //-------------------------------------------------------------------------
   /**
+   * Returns an instance with the rate updated.
+   * <p>
+   * This returns an {@code FxRate} with the same currency pair and an updated rate.
+   * For example, the operator could be used to add the forward points of an FX swap.
+   * <pre>
+   *   forwardRate = nearRate.mapRate(rate -> rate + forwardPoints);
+   * </pre>
+   *
+   * @param operator  the operator to be applied to the existing rate
+   * @return a new instance with the rate updated
+   */
+  public FxRate mapRate(DoubleUnaryOperator operator) {
```

Comment:
I have a very bad feeling about this method; it goes completely against the intent of `FxRate`. `FxRate` was very specifically designed not to expose the rate field to users.

If you have two currencies `AAA` and `BBB` and an `FxRate` which converts between them, you don't know whether the numeric rate is `AAA/BBB` or `BBB/AAA` and you don't need to care. If you want to convert an amount from `AAA` to `BBB` you call

    rate.convert(amount, Currency.AAA, Currency.BBB);

If you need the numeric value of the rate for converting from `AAA` to `BBB` you call

    rate.fxRate(Currency.AAA, Currency.BBB);

The logic to invert the rate when necessary is in `FxRate`. This spares the user from ever having to think about the direction of FX rates which is fiddly and error-prone.

This method exposes the underlying numerical rate and the map operation has a completely different meaning depending on whether the rate is `AAA/BBB` or `BBB/AAA`.

In order to make this method consistent with the rest of the class it should probably have this signature:

    FxRate mapRate(Currency fromCurrency, Currency toCurrency, DoubleUnaryOperator operator);

We would have to decide what the currency pair should be in the returned `FxRate`. The obvious choices are:

* The same as the pair in the receiver
* `fromCurrency/toCurrency`
* The market conventional pair